icu4c/source/i18n/measunit_extra.cpp

```diff
@@ -395,63 +435,89 @@ class Parser {
         return Token(match);
     }
 
+    /**
+     * Returns the next "single unit" via result.
+     *
+     * If a "-per-" was parsed, the result will have appropriate negative
+     * dimensionality.
+     *
+     * @param result Will be overwritten by the result, if status shows success.
+     * @param sawAnd If an "-and-" was parsed prior to finding the "single
+     * unit", sawAnd is set to true. If not, it is left as is.
+     * @param status ICU error code.
+     */
     void nextSingleUnit(SingleUnitImpl& result, bool& sawAnd, UErrorCode& status) {
-        sawAnd = false;
         if (U_FAILURE(status)) {
             return;
         }
 
-        if (!hasNext()) {
-            // probably "one"
-            return;
-        }
-
         // state:
         // 0 = no tokens seen yet (will accept power, SI prefix, or simple unit)
         // 1 = power token seen (will not accept another power token)
         // 2 = SI prefix token seen (will not accept a power or SI prefix token)
         int32_t state = 0;
 
-        // Maybe read a compound part
-        if (fIndex != 0) {
-            Token token = nextToken(status);
-            if (U_FAILURE(status)) {
-                return;
+        // Tracks the starting offset of the token currently under
+        // consideration.
+        int32_t previ = fIndex;
+        Token token = nextToken(status);
+        if (U_FAILURE(status)) { return; }
+
+        if (previ == 0) {
```

Comment:
I don't think you're actually using `previ` for anything.  Here, you can read `fIndex` instead of `previ`, and there isn't anywhere else in this function where you are reading the value from `previ`.

Reply:
Going with a bool: there's no default constructor for token, and I need the value of fIndex from before the nextToken call.